src/test/java/codeu/controller/ProfileServletTest.java

```diff
@@ -102,9 +119,109 @@ public void testDoPostUncleanedBio() throws IOException, ServletException {
-        Mockito.verify(mockResponse).sendRedirect("/users/candace");
+    }
+
+    @Test
+    public void testSendPalRequest() throws IOException, ServletException {
+        User userBob = new User(
+                UUID.randomUUID(), "Bob", "bobpassword", Instant.now(), false);
+        User userHelen = new User(
+                UUID.randomUUID(), "Helen", "helenpassword", Instant.now(), false);
+
+        UserStore mockUserStore = Mockito.mock(UserStore.class);
+        Mockito.when(mockUserStore.getUser("Bob")).thenReturn(userBob);
+        Mockito.when(mockUserStore.getUser("Helen")).thenReturn(userHelen);
+
+        Mockito.when(mockRequest.getRequestURI()).thenReturn("/users/Helen");
+        Mockito.when(mockRequest.getParameter("requestPal")).thenReturn("Bob");
+
+        Assert.assertEquals(0, userBob.getOutgoingRequests().size());
+        Assert.assertEquals(0, userHelen.getIncomingRequests().size());
+
+        profileServlet.setUserStore(mockUserStore);
+        profileServlet.sendPalRequest(mockRequest);
+
+        Assert.assertEquals(1, userBob.getOutgoingRequests().size());
+        Assert.assertEquals(1, userHelen.getIncomingRequests().size());
+    }
+
+    @Test
+    public void testAcceptPal() throws IOException, ServletException {
+        User userBob = new User(
+                UUID.randomUUID(), "Bob", "bobpassword", Instant.now(), false);
+        User userHelen = new User(
+                UUID.randomUUID(), "Helen", "helenpassword", Instant.now(), false);
+
+        UserStore mockUserStore = Mockito.mock(UserStore.class);
+        Mockito.when(mockUserStore.getUser("Bob")).thenReturn(userBob);
+        Mockito.when(mockUserStore.getUser("Helen")).thenReturn(userHelen);
+
+        Mockito.when(mockRequest.getRequestURI()).thenReturn("/users/Helen");
+        Mockito.when(mockRequest.getParameter("requestPal")).thenReturn("Bob");
+
+        Assert.assertEquals(0, userBob.getOutgoingRequests().size());
+        Assert.assertEquals(0, userHelen.getIncomingRequests().size());
+
+        profileServlet.setUserStore(mockUserStore);
+        profileServlet.sendPalRequest(mockRequest);
+
+        Assert.assertEquals(1, userBob.getOutgoingRequests().size());
+        Assert.assertEquals(1, userHelen.getIncomingRequests().size());
+
+        // Bob has sent Helen a pal request
+
+        Assert.assertEquals(0, userBob.getPals().size());
+        Assert.assertEquals(0, userHelen.getPals().size());
+
+        Mockito.when(mockRequest.getParameter("accept")).thenReturn("Bob");
+
+        profileServlet.acceptPal(mockRequest);
+
+        Assert.assertEquals(1, userBob.getPals().size());
+        Assert.assertEquals(1, userHelen.getPals().size());
+
+        Assert.assertEquals(0, userBob.getOutgoingRequests().size());
+        Assert.assertEquals(0, userHelen.getIncomingRequests().size());
+    }
+
+    @Test
+    public void testDeclinePal() throws IOException, ServletException {
+        User userBob = new User(
```

Comment:
Good tests .. you may want to move some of this common setup into a helper method